src/domain/pricing/bulk-editor/index.tsx

```diff
@@ -0,0 +1,1052 @@
+
+        return (
+          <div key={r.id} className="min-w-[314px]">
+            <PriceInput
+              amount={amount}
+              disabled={disabled}
+              currency={currencies[r.currency_code.toUpperCase()]}
+              hasVirtualFocus={isActive(variant.id, r.id)}
+              onMouseDown={(e) => onPriceInputClick(e, variant.id, r.id)}
+              onAmountChange={(a) => onAmountChange(variant.id, r.id, a)}
+              onKeyDown={(e) => onKeyDown(e, variant.id, r.id)}
+              /* prevent `onAmountChange` call from the library */
+              onBlur={(e) => e.preventDefault()}
+              data-id={getPriceKey(variant.id, r.id)}
+              className="js-bt-input"
+            />
+          </div>
+        )
+      })}
+    </div>
+  )
+}
+
+type PriceListBulkEditorProps = {
+  products: Product[]
+  activeRegions: Region[]
+  priceChanges: Record<string, string>
+  isVariantInPriceList: (variantId: string) => boolean
+  setPriceChanges: (a: Record<string, string>) => void
+}
+
+/**
+ * Root component for the bulk editor.
+ * Implements table rendering and all multiedit/multiselect behaviour.
+ */
+function PriceListBulkEditor(props: PriceListBulkEditorProps) {
+  const {
+    activeRegions,
+    products,
+    priceChanges,
+    setPriceChanges,
+    isVariantInPriceList,
+  } = props
+
+  const [activeFields, setActiveFields] = useState({})
+  const [currentEditAmount, setCurrentEditAmount] = useState<
+    string | undefined
+  >()
+
+  const { current: pointers } = useRef<CellPointers>({} as CellPointers)
+
+  /**
+   * A struct for mapping between table cell indexes and variant/regions ids.
+   */
+  const matrix = useMemo(() => {
+    const variants = products.reduce(
+      (acc, p) => [...acc, ...p.variants.map((v) => v.id)],
+      []
+    )
+
+    return [variants, activeRegions.map((r) => r.id)]
+  }, [activeRegions, products])
+
+  useEffect(() => {
+    const handler = (e) => {
+      const isPriceInputClicked = e.target.classList.contains("js-bt-input")
+
+      // artificial blur event
+      if (!isPriceInputClicked) {
+        setCurrentEditAmount(undefined)
+        setActiveFields({})
+      }
+    }
+
+    document.addEventListener("mousedown", handler)
+    return () => document.removeEventListener("mousedown", handler)
+  }, [])
+
+  /* ********** HELPERS ********** */
+
+  const activateField = (cellKey: string) => {
+    setActiveFields({ ...activeFields, [cellKey]: true })
+  }
+
+  const deactivateField = (cellKey: string) => {
+    const tmp = { ...activeFields }
+    delete tmp[cellKey]
+
+    setActiveFields(tmp)
+  }
+
+  const toggleActive = (cellKey: string) => {
+    if (activeFields[cellKey]) {
+      deactivateField(cellKey)
+    } else {
+      activateField(cellKey)
+    }
+  }
+
+  const setActiveCellsFromPointers = () => {
+    const next = {}
+    const [variants, regions] = matrix
+
+    for (let i = pointers.minV; i <= pointers.maxV; i++) {
```

Comment:
I was wondering what minV maxV and the same for R was standing for. now I know, would probably be worth naming them differently